src/target/riscv/riscv.c

```diff
@@ -1614,6 +1614,282 @@ static int riscv_hit_trigger_hit_bit(struct target *target, uint32_t *unique_id)
 	return ERROR_OK;
 }
 
+static inline uint32_t get_bits_field(riscv_insn_t b, uint32_t lo, uint32_t len) {
```

Comment:
This function is needed to extract individual bits (for imm) that do not have a mask in the file encoding.h from the instruction.
Example for C.LW:
Now
```
static inline int16_t get_imm_clw(riscv_insn_t instruction) {
	return (get_bits_field(instruction, 6, 1) << 2)
	       + (get_bits_field(instruction, 10, 3) << 3)
		   + (get_bits_field(instruction, 5, 1) << 6);
}
```
Or if use func get_field32()
```
static inline int16_t get_imm_clw(riscv_insn_t instruction) {
	uint32_t imm2i6 = get_field32(instruction, INSN_FIELD_C_UIMM2); // similar mask but this is imm offset[2|6]
	return ((imm2i6 & 0x2) << 2)
		   | (get_field32(instruction, INSN_FIELD_C_UIMM7HI) << 3) // similar mask but this is imm offset[5:3]
		   | ((imm2i6 & 0x1) << 6);
}
```

Reply:
> This function is needed to extract individual bits (for imm) that do not have a mask in the file encoding.h from the instruction. Example for C.LW: Now
> 
> ```
> static inline int16_t get_imm_clw(riscv_insn_t instruction) {
> 	return (get_bits_field(instruction, 6, 1) << 2)
> 	       + (get_bits_field(instruction, 10, 3) << 3)
> 		   + (get_bits_field(instruction, 5, 1) << 6);
> }
> ```
> 
> Or if use func get_field32()
> 
> ```
> static inline int16_t get_imm_clw(riscv_insn_t instruction) {
> 	uint32_t imm2i6 = get_field32(instruction, INSN_FIELD_C_UIMM2); // similar mask but this is imm offset[2|6]
> 	return ((imm2i6 & 0x2) << 2)
> 		   | (get_field32(instruction, INSN_FIELD_C_UIMM7HI) << 3) // similar mask but this is imm offset[5:3]
> 		   | ((imm2i6 & 0x1) << 6);
> }
> ```

Does that perhaps suggest that `encoding.h` might need updating rather than creating additional "local" utility functions?

Even if it doesn't, it probably makes sense to use descriptive `#define`s rather than "magic numbers" for the mask/shift values in order to aid readability?

Reply:
> Does that perhaps suggest that `encoding.h` might need updating rather than creating additional "local" utility functions?
> 
> Even if it doesn't, it probably makes sense to use descriptive `#define`s rather than "magic numbers" for the mask/shift values in order to aid readability?

I understand, I'll look at how these fields are described in the riscv-isa-manual itself and try to change my code.

Reply:
Just to note, if changes to `encoding.h` are merited then I think that these would probably need to be done upstream where this file is mastered and auto-generated.

* https://github.com/riscv-collab/riscv-openocd/blob/ca7d88252633a5d452ee92e6f8f3eb20cadfb633/src/target/riscv/encoding.h#L5-L8

Reply:
I want to do this.
Does this bits magic look better?
```
// c.lw rd_p rs1_p c_uimm7lo c_uimm7hi     1..0=0 15..13=2
// latex_mapping['c_uimm7lo'] = 'uimm[2$\\vert$6]'
// latex_mapping['c_uimm7hi'] = 'uimm[5:3]'
static inline uint16_t get_imm_clw(riscv_insn_t instruction) {
    uint32_t uimm_2and6_bits = get_field32(instruction, INSN_FIELD_C_UIMM7LO);
    uint32_t uimm_3to5_bits = get_field32(instruction, INSN_FIELD_C_UIMM7HI);
    return ((uimm_2and6_bits & 0x2) << 2) | (uimm_3to5_bits << 3)
           | ((uimm_2and6_bits & 0x1) << 6);
}
```